src/react/ReactElementSet.js

```diff
@@ -10,171 +10,41 @@
 /* @flow */
 
 import { Realm } from "../realm.js";
-import {
-  AbstractValue,
-  ArrayValue,
-  FunctionValue,
-  NumberValue,
-  ObjectValue,
-  StringValue,
-  SymbolValue,
-  Value,
-} from "../values/index.js";
+import { ObjectValue, Value } from "../values/index.js";
 import invariant from "../invariant.js";
-import { isReactElement, getProperty } from "./utils";
-import { HashSet } from "../methods/index.js";
+import { type ReactSetKeyMap, ReactSet } from "./ReactSet.js";
+import { ResidualReactElementVisitor } from "../serializer/ResidualReactElementVisitor.js";
 
-type ReactElementValueMapKey = Value | number | string;
-type ReactElementValueMap = Map<ReactElementValueMapKey, ReactElementNode>;
-
-type ReactElementKeyMapKey = string | number | SymbolValue;
-type ReactElementKeyMap = Map<ReactElementKeyMapKey, ReactElementValueMap>;
-
-type ReactElementNode = {
-  map: ReactElementKeyMap,
-  value: ObjectValue | ArrayValue | null,
-};
-
-// ReactElementSet keeps records around of the values
-// of ReactElement/JSX nodes so we can return the same immutable values
-// where possible, i.e. <div /> === <div />
-//
-// Rather than uses hashes, this class uses linked Maps to track equality of objects.
-// It does this by recursively iterating through objects, by their properties/symbols and using
-// each property key as a map, and then from that map, each value as a map. The value
-// then links to the subsequent property/symbol in the object. This approach ensures insertion
-// is maintained through all objects.
-
-export default class ReactElementSet {
-  constructor(realm: Realm, equivalenceSet: HashSet<AbstractValue>) {
-    this.realm = realm;
-    this.equivalenceSet = equivalenceSet;
+export class ReactElementSet extends ReactSet {
```

Comment:
Does inheritance really help here?

I know we use it for the value taxonomy but I'd prefer to avoid it for the rest of the logic. I imagine you could instantiate `ReactSet` instead, keep it as a field, and delegate to it when necessary.

Reply:
What's wrong with inheritance?

Reply:
Same reason we don't use it in components? :-)

Inheritance makes the dependencies between parts very implicit. It's hard for me to tell what is the contract between `ReactSet` and its subclasses. Do they call its methods? Does it call their methods? Does both happen? Since all interaction happens through `this.something()` I have to check every method name to verify where it's defined.

This problem usually becomes even more difficult with time, as superclass become coupled to subclasses, and then one of them needs to behave differently but can't "break out".

Using delegation would make it clearer what `ReactSet` is doing, and how each of the specific "sets" uses this functionality.

Reply:
I have to disagree here, This isn't the same as UI components, this is business logic. I actually find inheritance is fine in cases like this when there's only a single level of inheritance and the contract is explicit.

I'll refactor the classes again though as I don't want to block this PR on arguments of the validity of inheritance :D

Reply:
It's not blocking, I'm just finding it harder to understand what the code is doing because of the implicitness.

Reply:
I've redone it all and removed the classes, it's all functions and argument passing now. Functional FTW